storage/pruning/triePruningStorer.go

```diff
@@ -0,0 +1,194 @@
+	var persisters []*persisterData
+	persistersMapByEpoch := make(map[uint32]*persisterData)
+
+	closeOldPersisters := false
+	for epoch := int64(args.StartingEpoch); epoch >= oldestEpochKeep; epoch-- {
+		log.Debug("initTriePersisterInEpoch(): createPersisterDataForEpoch", "identifier", args.Identifier, "epoch", epoch, "shardID", shardIDStr)
+		p, err := createPersisterDataForEpoch(args, uint32(epoch), shardIDStr)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		persistersMapByEpoch[uint32(epoch)] = p
+
+		if epoch < oldestEpochActive && closeOldPersisters {
+			err = p.Close()
+			if err != nil {
+				log.Debug("persister.Close()", "identifier", args.Identifier, "error", err.Error())
+			}
+		} else {
+			persisters = append(persisters, p)
+			log.Debug("appended a pruning active persister", "epoch", epoch, "identifier", args.Identifier)
+		}
+
+		if !closeOldPersisters {
+			closeOldPersisters = shouldCloseOldPersisters(p)
+		}
+	}
+
+	return persisters, persistersMapByEpoch, nil
+}
+
+func shouldCloseOldPersisters(pd *persisterData) bool {
+	val, err := pd.persister.Get([]byte(common.ActiveDBKey))
+	if err != nil {
+		return false
+	}
+
+	if bytes.Equal(val, []byte(common.ActiveDBVal)) {
+		return true
+	}
+
+	return false
+}
+
+// PutInEpochWithoutCache adds data to persistence medium related to the specified epoch and updates the bloom filter
+func (ps *triePruningStorer) PutInEpochWithoutCache(key []byte, data []byte, epoch uint32) error {
+	ps.lock.RLock()
+	pd, exists := ps.persistersMapByEpoch[epoch]
+	ps.lock.RUnlock()
+	if !exists {
+		return fmt.Errorf("put in epoch: persister for epoch %d not found", epoch)
+	}
+
+	persister, closePersister, err := ps.createAndInitPersisterIfClosedProtected(pd)
+	if err != nil {
+		return err
+	}
+	defer closePersister()
+
+	err = persister.Put(key, data)
+	if err != nil {
+		return err
+	}
+
+	if ps.bloomFilter != nil {
+		ps.bloomFilter.Add(key)
+	}
+
+	return nil
+}
+
+// GetFromOldEpochsWithoutAddingToCache searches the old epochs for the given key without adding to the cache
+func (ps *triePruningStorer) GetFromOldEpochsWithoutAddingToCache(key []byte) ([]byte, error) {
+	v, ok := ps.cacher.Get(key)
+	if ok {
+		return v.([]byte), nil
+	}
+
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	if ps.bloomFilter != nil && !ps.bloomFilter.MayContain(key) {
+		return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
+	}
+
+	for idx := uint32(1); (idx < ps.numOfActivePersisters) && (idx < uint32(len(ps.activePersisters))); idx++ {
+		val, err := ps.activePersisters[idx].persister.Get(key)
+		if err != nil {
+			continue
+		}
+
+		return val, nil
+	}
+
+	return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
+}
+
+// GetFromLastEpoch searches only the last epoch storer for the given key
+func (ps *triePruningStorer) GetFromLastEpoch(key []byte) ([]byte, error) {
```

Comment:
Last epoch as in not the current, the one exactly before?

Reply:
Yes. This is needed because in the case of a trie snapshot, all the values from the previous db need to be added to the current db.

Reply:
ok